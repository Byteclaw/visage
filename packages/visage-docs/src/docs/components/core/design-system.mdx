---
title: 'Design system component'
components: ['DesignSystem']
---

# Design system

Every application that wants to use Visage as it's design system must be wrapped in DesignSystem component.

DesignSystem also provides a [LayerManager](/components/layer-manager) for zIndex management and [ToastManager](/components/toast-manager) for [Toast](/components/feedback/toast) components.

## Import

```js transpile=false
import { DesignSystem } from '@byteclaw/visage';
```

## Usage

```js transpile=false
<ResponsiveDesignSystem theme={theme}>
  <Heading>Hello world</Heading>
<ResponsiveDesignSystem>
```

## Prop Types

<PropTypes component="DesignSystem" />
