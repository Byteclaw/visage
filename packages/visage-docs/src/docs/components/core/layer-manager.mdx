---
title: 'LayerManager component'
components: ['LayerManager']
---

# LayerManager

LayerManager is responsible for managing zIndexes accross your application. Every new layer must be wrapped in LayerManager in order to guarantee that zIndexes are correct.

Layer manager exposes hook [useLayerManager](/hooks/use-layer-manager) which can be used to get the current zIndex and nesting level information from LayerManager.

## Import

```js transpile=false
import { LayerManager } from '@byteclaw/visage';
```

## Usage

```js transpile=false
function Dialog() {
  const { zIndex } = useLayerManager();

  return <Box styles={{ position: 'absolute', zIndex }}>This is new layer</Box>;
}

<LayerManager>
  <Dialog />
</LayerManager>;
```

## Props

### Optional

#### increaseBy

- **number**

Telss the layer manager about the number that is added to current zIndex if nesting level increases.

```js transpile=false
<LayerManager incraseBy={100}>
  zIndex will be 200 because default zIndex from root is 100
  <LayerManager>
    zIndex will be 300 (because increaseBy is taken from context)
  </LayerManager>
</LayerManager>
```

## Prop Types

<PropTypes component="ToastManager" />

## Used by

- [DesignSystem](/components/design-system)
- [Drawer](/components/overlays/drawer)
- [Modal](/components/overlays/modal)
- [Popover](/components/overlays/popover)
- [ResponsiveDesignSystem](/components/responsive-design-system)
