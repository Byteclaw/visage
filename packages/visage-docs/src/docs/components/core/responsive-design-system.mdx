---
title: 'Responsive design system component'
components: ['ResponsiveDesignSystem']
---

# Responsive design system

`ResponsiveDesignSystem` acts as a shortcut for `<DesignSystem responsive />`. It accepts the same props as [DesignSystem](/components/core/design-system) component.

## Import

```js transpile=false
import { ResponsiveDesignSystem } from '@byteclaw/visage';
```

## Usage

```js transpile=false
<ResponsiveDesignSystem theme={theme}>
  <Heading>Hello world</Heading>
<ResponsiveDesignSystem>
```

## Prop Types

<PropTypes component="ResponsiveDesignSystem" />
