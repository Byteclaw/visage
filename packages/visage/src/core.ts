import {
  ComponentFactory,
  createComponent as baseCreateComponent,
} from '@byteclaw/visage-core';
import { StyleProps } from './types';

export const createComponent: ComponentFactory<StyleProps> = baseCreateComponent;
