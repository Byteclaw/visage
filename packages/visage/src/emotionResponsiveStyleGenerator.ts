import {
  resolveStyleSheets,
  StyleGenerator,
  ResolvedStyleSheet,
} from '@byteclaw/visage-core';
import { css } from 'emotion';

/**
 * Generates styles with media queries
 */
export function createResponsiveEmotionStyleGenerator(
  /**
   * Media queries sorted from smallest to largest device
   *
   * Please provide only conditions e.g. only screen or screen and (min-width: 40em)
   */
  mediaQueries: string[],
): StyleGenerator {
  return (styleSheets, ctx) => {
    const sheet: { [mediaQuery: string]: ResolvedStyleSheet } = {};

    // iterate through media queries and resolve style for them
    mediaQueries.forEach((mediaQuery, breakpoint) => {
      sheet[`@media ${mediaQuery}`] = resolveStyleSheets(styleSheets, {
        ...ctx,
        breakpoint,
      });
    });

    return {
      className: css(sheet),
    };
  };
}
