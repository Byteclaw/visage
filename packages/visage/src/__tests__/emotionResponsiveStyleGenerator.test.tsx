import { render } from '@testing-library/react';
import React from 'react';
import { TestDesignSystem } from './TestDesignSystem';
import { createResponsiveEmotionStyleGenerator } from '../emotionResponsiveStyleGenerator';
import { createComponent } from '../core';

const MOBILE_BP = `only screen`; // 40em
const TABLET_BP = `screen and (min-width: ${641 / 16}em)`; // 40.0625em
const DESKTOP_BP = `screen and (min-width: ${1025 / 16}em)`; // 64.036em

const defaultBreakpoints = [MOBILE_BP, TABLET_BP, DESKTOP_BP];

const styleGenerator = createResponsiveEmotionStyleGenerator(
  defaultBreakpoints,
);

describe('emotion responsive style generator', () => {
  it('works correctly', () => {
    const Box = createComponent('div', {
      styles: {
        color: 'primary',
        fontFamily: 'body',
        fontSize: [1, 2, 3],
        margin: [2, 3, 4],
        '@media only screen': {
          color: 'blue',
          fontSize: 1,
        },
      },
    });

    const { asFragment, rerender } = render(
      <TestDesignSystem is={0} styleGenerator={styleGenerator}>
        <Box />{' '}
      </TestDesignSystem>,
    );

    expect(asFragment()).toMatchSnapshot();

    rerender(
      <TestDesignSystem is={1} styleGenerator={styleGenerator}>
        <Box />{' '}
      </TestDesignSystem>,
    );

    expect(asFragment()).toMatchSnapshot();

    rerender(
      <TestDesignSystem is={2} styleGenerator={styleGenerator}>
        <Box />{' '}
      </TestDesignSystem>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
