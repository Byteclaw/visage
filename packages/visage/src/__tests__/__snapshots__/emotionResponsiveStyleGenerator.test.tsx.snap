// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emotion responsive style generator works correctly 1`] = `
<DocumentFragment>
  @media only screen {
  .emotion-0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 600;
  }
}

@media screen and (min-width:40.0625em) {
  .emotion-0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 600;
  }
}

@media screen and (min-width:64.0625em) {
  .emotion-0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 600;
  }
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  @media only screen {
  .emotion-0 {
    color: rgb(173,216,230);
    font-family: sans-serif;
    font-size: 26px;
    margin: 16px;
  }

@media only screen {
    .emotion-0 {
      color: rgb(0,0,255);
      font-size: 26px;
    }
}
}

@media screen and (min-width:40.0625em) {
  .emotion-0 {
    color: rgb(173,216,230);
    font-family: sans-serif;
    font-size: 41px;
    margin: 24px;
  }

@media only screen {
    .emotion-0 {
      color: rgb(0,0,255);
      font-size: 26px;
    }
}
}

@media screen and (min-width:64.0625em) {
  .emotion-0 {
    color: rgb(173,216,230);
    font-family: sans-serif;
    font-size: 66px;
    margin: 32px;
  }

@media only screen {
    .emotion-0 {
      color: rgb(0,0,255);
      font-size: 26px;
    }
}
}

<div
    class="emotion-0"
  />
   
</DocumentFragment>
`;

exports[`emotion responsive style generator works correctly 2`] = `
<DocumentFragment>
  @media only screen {
  .emotion-0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 600;
  }
}

@media screen and (min-width:40.0625em) {
  .emotion-0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 600;
  }
}

@media screen and (min-width:64.0625em) {
  .emotion-0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 600;
  }
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  @media only screen {
  .emotion-0 {
    color: rgb(173,216,230);
    font-family: sans-serif;
    font-size: 26px;
    margin: 16px;
  }

@media only screen {
    .emotion-0 {
      color: rgb(0,0,255);
      font-size: 26px;
    }
}
}

@media screen and (min-width:40.0625em) {
  .emotion-0 {
    color: rgb(173,216,230);
    font-family: sans-serif;
    font-size: 41px;
    margin: 24px;
  }

@media only screen {
    .emotion-0 {
      color: rgb(0,0,255);
      font-size: 26px;
    }
}
}

@media screen and (min-width:64.0625em) {
  .emotion-0 {
    color: rgb(173,216,230);
    font-family: sans-serif;
    font-size: 66px;
    margin: 32px;
  }

@media only screen {
    .emotion-0 {
      color: rgb(0,0,255);
      font-size: 26px;
    }
}
}

<div
    class="emotion-0"
  />
   
</DocumentFragment>
`;

exports[`emotion responsive style generator works correctly 3`] = `
<DocumentFragment>
  @media only screen {
  .emotion-0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 600;
  }
}

@media screen and (min-width:40.0625em) {
  .emotion-0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 600;
  }
}

@media screen and (min-width:64.0625em) {
  .emotion-0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 600;
  }
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  @media only screen {
  .emotion-0 {
    color: rgb(173,216,230);
    font-family: sans-serif;
    font-size: 26px;
    margin: 16px;
  }

@media only screen {
    .emotion-0 {
      color: rgb(0,0,255);
      font-size: 26px;
    }
}
}

@media screen and (min-width:40.0625em) {
  .emotion-0 {
    color: rgb(173,216,230);
    font-family: sans-serif;
    font-size: 41px;
    margin: 24px;
  }

@media only screen {
    .emotion-0 {
      color: rgb(0,0,255);
      font-size: 26px;
    }
}
}

@media screen and (min-width:64.0625em) {
  .emotion-0 {
    color: rgb(173,216,230);
    font-family: sans-serif;
    font-size: 66px;
    margin: 32px;
  }

@media only screen {
    .emotion-0 {
      color: rgb(0,0,255);
      font-size: 26px;
    }
}
}

<div
    class="emotion-0"
  />
   
</DocumentFragment>
`;
